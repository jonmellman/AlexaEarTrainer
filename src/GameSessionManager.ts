import * as Alexa from 'ask-sdk-core'
import {
	GameSession
} from './game'

// Wrap AttribuetsManager so we can type the attributes as GameSession
export class GameSessionManager {
	constructor(readonly handlerInput: Alexa.HandlerInput) { }

	public getSession(): GameSession {
		return this.handlerInput.attributesManager.getSessionAttributes()
	}

	public setSession(gameSession: GameSession): void {
		return this.handlerInput.attributesManager.setSessionAttributes(gameSession)
	}
}
