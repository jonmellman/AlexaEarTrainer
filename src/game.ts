// Nothing Alexa-specific in this file.

// https://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies
export enum Note {
	C4 = 60,
	D4 = 62,
	E4 = 64,
	F4 = 65,
	G4 = 67,
	A4 = 69,
	B4 = 71,
	C5 = 72
}

export enum Key {
	C = 0,
	'C#',
	D,
	Eb,
	E,
	F,
	'F#',
	G,
	Ab,
	A,
	Bb,
	B,
}

export enum Interval {
	ROOT = 0,
	MINOR_SECOND,
	MAJOR_SECOND,
	MINOR_THIRD,
	MAJOR_THIRD,
	PERFECT_FOURTH,
	TRITONE,
	PERFECT_FIFTH,
	MINOR_SIXTH,
	MAJOR_SIXTH,
	MINOR_SEVENTH,
	MAJOR_SEVENTH,
	OCTAVE
}

const Quality = {
	MAJOR: [
		Interval.ROOT,
		Interval.MAJOR_SECOND,
		Interval.MAJOR_THIRD,
		Interval.PERFECT_FOURTH,
		Interval.PERFECT_FIFTH,
		Interval.MAJOR_SIXTH,
		Interval.MAJOR_SEVENTH,
		Interval.OCTAVE
	]
}

interface Level {
	description: string,
	key: Key,
	targetIntervals: Interval[],
	numRounds: number
}

interface Stat {
	guess: CurrentRound['targetInterval'],
	answer: CurrentRound['targetInterval']
}

export type GameSession = LevelInProgress | LevelComplete

interface LevelInProgress {
	state: 'LEVEL_IN_PROGRESS',
	level: number,
	stats: Stat[],
	currentRound: CurrentRound
}

interface LevelComplete {
	state: 'LEVEL_COMPLETE',
	level: number,
	stats: Stat[]
}

export interface CurrentRound {
	roundNumber: number
	key: Key
	targetInterval: Interval
}



export const levels: Level[] = [{
	description: 'First half of the C Major scale',
	key: Key.C,
	targetIntervals: Quality.MAJOR.slice(0, Quality.MAJOR.indexOf(Interval.PERFECT_FIFTH)),
	numRounds: 5
}, {
	description: 'Second half of the C Major scale',
	key: Key.C,
	targetIntervals: Quality.MAJOR.slice(Quality.MAJOR.indexOf(Interval.PERFECT_FIFTH)),
	numRounds: 5
}, {
	description: 'Full octave of the C Major scale',
	key: Key.C,
	targetIntervals: Quality.MAJOR,
	numRounds: 5
}]

// Given a game level, select two pitches
// Given a pitch, return the audio file name
// Given two pitches, calculate the distance
// Given the distance, return the correct answer(s)


// Level 1: (C4), (D4, E4, F4)
// Level 2: (C4), (G4, A4, B4)
// Level 3: (C4), (C4, D4, E4, F4, G4, A4, B4)

export const getNewGame = (level: GameSession['level'] = 1): LevelInProgress => {
	return {
		state: 'LEVEL_IN_PROGRESS',
		level,
		stats: [],
		currentRound: getNewRound(level)
	}
}

export const evaluateGuess = (gameSession: LevelInProgress, guessInterval: Interval): LevelInProgress | LevelComplete => {
	const isLevelComplete = gameSession.currentRound.roundNumber === levels[gameSession.level - 1].numRounds
	return isLevelComplete ? levelComplete(gameSession, guessInterval) : levelProgress(gameSession, guessInterval)

	function levelComplete(gameSession: LevelInProgress, guessInterval: Interval): LevelComplete {
		return {
			...gameSession,
			state: 'LEVEL_COMPLETE',
			stats: [...gameSession.stats, {
				guess: guessInterval,
				answer: gameSession.currentRound.targetInterval
			}],
		}
	}

	function levelProgress(gameSession: LevelInProgress, guessInterval: Interval): LevelInProgress {
		return {
			...gameSession,
			stats: [...gameSession.stats, {
				guess: guessInterval,
				answer: gameSession.currentRound.targetInterval
			}],
			currentRound: getNewRound(gameSession.level, gameSession.currentRound.roundNumber)
		}
	}
}

function getNewRound(level: GameSession['level'], previousRoundNumber: CurrentRound['roundNumber'] = 0): CurrentRound {
	return {
		roundNumber: previousRoundNumber + 1,
		...getIntervalsForLevel(level)
	}

	function getIntervalsForLevel(levelNumber: number) {
		const level = levels[levelNumber - 1]

		if (!level) {
			throw new Error(`No level configuration for level ${levelNumber}!`)
		}

		return {
			key: level.key,
			targetInterval: getRandomElement(level.targetIntervals)
		}

		function getRandomElement<T>(elements: Array<T>): T {
			return elements[Math.floor(Math.random() * elements.length)];
		}
	}
}
