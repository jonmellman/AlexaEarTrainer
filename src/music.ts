// https://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies
export enum Note {
	// TODO
	C4 = 60,
	D4 = 62,
	E4 = 64,
	F4 = 65,
	G4 = 67,
	A4 = 69,
	B4 = 71,
	C5 = 72
	// TODO
}

export enum Key {
	C = 0,
	'C#',
	D,
	Eb,
	E,
	F,
	'F#',
	G,
	Ab,
	A,
	Bb,
	B,
}

export enum Interval {
	ROOT = 0,
	MINOR_SECOND,
	MAJOR_SECOND,
	MINOR_THIRD,
	MAJOR_THIRD,
	PERFECT_FOURTH,
	TRITONE,
	PERFECT_FIFTH,
	MINOR_SIXTH,
	MAJOR_SIXTH,
	MINOR_SEVENTH,
	MAJOR_SEVENTH,
	OCTAVE
}

export const Quality = {
	MAJOR: [
		Interval.ROOT,
		Interval.MAJOR_SECOND,
		Interval.MAJOR_THIRD,
		Interval.PERFECT_FOURTH,
		Interval.PERFECT_FIFTH,
		Interval.MAJOR_SIXTH,
		Interval.MAJOR_SEVENTH,
		Interval.OCTAVE
	]
}

export const getNoteFromKeyAtInterval = (key: Key, interval: Interval): Note => {
	const note = (key + interval) + 60 // MIDI offset

	if (!Note[note]) {
		throw new Error(`Unsupported note: ${note}. Key was ${Key[key]} and Interval was ${Interval[interval]}`)
	}

	return note
}
