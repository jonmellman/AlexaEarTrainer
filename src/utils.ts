import * as Alexa from 'ask-sdk-core'
import {
	GameState
} from './game/state'

// Wrap AttribuetsManager as a new utility so we can type the attributes as GameSession.
export class GameSessionManager {
	constructor(readonly handlerInput: Alexa.HandlerInput) { }

	public getSession(): GameState {
		return this.handlerInput.attributesManager.getSessionAttributes()
	}

	public setSession(gameSession: GameState): void {
		return this.handlerInput.attributesManager.setSessionAttributes(gameSession)
	}
}

export const isTestEnvironment = (): boolean => process.env.JEST_WORKER_ID !== undefined;
