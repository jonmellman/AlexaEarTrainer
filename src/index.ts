import * as Alexa from 'ask-sdk-core'
import {
	RequestHandler,
} from 'ask-sdk-core';

import {
	getNewGame,
	evaluateGuess,
} from './game'
import { GameSessionManager } from './GameSessionManager'
import * as speech from './speech';

const AnswerHandler: RequestHandler = {
	canHandle(handlerInput) {
		return !Alexa.isNewSession(handlerInput.requestEnvelope) &&
			Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest' &&
			Alexa.getIntentName(handlerInput.requestEnvelope) === 'AnswerIntent';
	},
	async handle(handlerInput) {
		const gameSessionManager = new GameSessionManager(handlerInput)

		const intervalDistanceGuess = extractIntervalDistanceFromAnswer(handlerInput)
		const gameSession = evaluateGuess(gameSessionManager.getSession(), intervalDistanceGuess)

		const stat = gameSession.stats[gameSession.stats.length - 1]
		const isCorrect = stat.guess === stat.answer

		const counts = {
			correct: gameSession.stats.filter(stat => stat.guess === stat.answer).length,
			total: gameSession.stats.length
		}

		gameSessionManager.setSession(gameSession)

		return handlerInput.responseBuilder
			.speak(speech.compose(
				speech.assess(isCorrect),
				(gameSession.currentRound ? speech.question(gameSession.currentRound) : speech.roundComplete(counts.correct, counts.total, gameSession.level + 1))
			))
			.withShouldEndSession(false)
			.getResponse();

		function extractIntervalDistanceFromAnswer(handlerInput: Alexa.HandlerInput): number {
			const slot = Alexa.getSlot(handlerInput.requestEnvelope, 'intervalName') // `intervalName` is the literal name of the slot.

			if (!slot?.resolutions?.resolutionsPerAuthority) {
				throw new Error(`No slot resolutions found on '${slot}'`)
			}

			const resolutionsPerAuthority = slot.resolutions.resolutionsPerAuthority

			if (!resolutionsPerAuthority[0].values) {
				throw new Error(`'${slot.value}' didn't match '${slot.name}' slots`)
			}

			/*
				If we created our slots correctly in the Alexa Developer Console, we should have an ID that corresponds to the interval distance.
				If the guess is "minor second", then the distance will be 1. "major second" will be 2, etc.
			*/
			const intervalDistanceString = resolutionsPerAuthority[0].values[0].value.id
			const intervalDistance = parseInt(intervalDistanceString)

			if (isNaN(intervalDistance)) {
				throw new Error(`Error converting '${intervalDistanceString}' to a number`)
			}

			return intervalDistance
		}
	}
}

const LaunchRequest: RequestHandler = {
	canHandle(handlerInput) {
		return Alexa.isNewSession(handlerInput.requestEnvelope) ||
			Alexa.getRequestType(handlerInput.requestEnvelope) === 'LaunchRequest';
	},
	async handle(handlerInput) {
		const gameSessionManager = new GameSessionManager(handlerInput)
		const level = 1

		// Set state for new game
		gameSessionManager.setSession(getNewGame(level))

		// Return UI for new game
		const currentRound = gameSessionManager.getSession().currentRound

		if (!currentRound) {
			throw new Error('Expected currentRound when starting a new game')
		}

		return handlerInput.responseBuilder
			.speak(speech.compose(
				speech.welcome(),
				speech.levelIntroduction(level),
				speech.question(currentRound)
			))
			.reprompt('Bee boop') // TODO
			.withShouldEndSession(false)
			.getResponse();
	},
};

const FallbackHandler: Alexa.RequestHandler = {
	canHandle(handlerInput) {
		if (isTestEnvironment()) {
			throw new Error(`Fallback handler invoked request: ${Alexa.getRequestType(handlerInput.requestEnvelope)}, intent: ${Alexa.getIntentName(handlerInput.requestEnvelope)}!`)
		}

		return true
	},
	handle(handlerInput) {
		// TODO: implement
		return handlerInput.responseBuilder
			.speak(speech.compose(
				'Fallback handler',
				Alexa.getRequestType(handlerInput.requestEnvelope),
				Alexa.getIntentName(handlerInput.requestEnvelope)
			))
			.getResponse()
	}
}

const ErrorHandler: Alexa.ErrorHandler = {
	canHandle() {
		// Unhandled errors in tests are easier to debug
		return !isTestEnvironment();
	},
	handle(handlerInput, error) {
		console.log(`Error handled: ${error.stack}`)

		const speechOutput = error.stack || 'No stack trace'
		const repromptText = speechOutput;

		return handlerInput.responseBuilder
			.speak(speechOutput)
			.reprompt(repromptText)
			.getResponse();
	},
};

const SessionEndedRequestHandler: Alexa.RequestHandler = {
	canHandle(handlerInput) {
		return handlerInput.requestEnvelope.request.type === 'SessionEndedRequest';
	},
	handle(handlerInput) {
		// TODO
		// @ts-ignore
		console.log(`Session ended with reason: ${handlerInput.requestEnvelope.request.reason}`);

		return handlerInput.responseBuilder.getResponse();
	},
};

const NextLevelHandler: Alexa.RequestHandler = {
	canHandle(handlerInput) {
		return !Alexa.isNewSession(handlerInput.requestEnvelope) &&

			// Round is over
			new GameSessionManager(handlerInput).getSession().currentRound === undefined &&

			// User said yes
			Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest' &&
			Alexa.getIntentName(handlerInput.requestEnvelope) === 'AMAZON.YesIntent';
	},
	handle(handlerInput) {
		const gameSessionManager = new GameSessionManager(handlerInput)
		const level = gameSessionManager.getSession().level + 1

		// Set state for new game
		gameSessionManager.setSession(getNewGame(level))

		// Return UI for new game
		const currentRound = gameSessionManager.getSession().currentRound

		if (!currentRound) {
			throw new Error('Expected currentRound when starting a new game')
		}

		return handlerInput.responseBuilder
			.speak(speech.compose(
				speech.levelIntroduction(level),
				speech.question(currentRound)
			))
			.withShouldEndSession(false)
			.getResponse();
	},
}

const StopHandler: Alexa.RequestHandler = {
	canHandle(handlerInput) {
		return (
				// User said Stop or Cancel
				Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest' &&
				(
					Alexa.getIntentName(handlerInput.requestEnvelope) === 'AMAZON.StopIntent' ||
					Alexa.getIntentName(handlerInput.requestEnvelope) === 'AMAZON.CancelIntent'
				)
			) || (
				// Round is over
				!Alexa.isNewSession(handlerInput.requestEnvelope) &&
				new GameSessionManager(handlerInput).getSession().currentRound === undefined &&

				// User said No to continuing
				Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest' &&
				Alexa.getIntentName(handlerInput.requestEnvelope) === 'AMAZON.NoIntent'
			)
	},
	handle(handlerInput) {
		return handlerInput.responseBuilder
			.speak('Thanks for playing!')
			.withShouldEndSession(true)
			.getResponse();
	},
}

const RepeatQuestionHandler: Alexa.RequestHandler = {
	canHandle(handlerInput) {
		return !Alexa.isNewSession(handlerInput.requestEnvelope) &&
			new GameSessionManager(handlerInput).getSession().currentRound !== undefined &&
			Alexa.getRequestType(handlerInput.requestEnvelope) === 'IntentRequest' &&
			Alexa.getIntentName(handlerInput.requestEnvelope) === 'AMAZON.RepeatIntent'
	},
	handle(handlerInput) {
		const gameSessionManager = new GameSessionManager(handlerInput)
		const gameSession = gameSessionManager.getSession()

		if (!gameSession.currentRound) {
			throw new Error('Asked to repeat a question, but no currentRound!')
		}

		return handlerInput.responseBuilder
			.speak(speech.question(gameSession.currentRound))
			.withShouldEndSession(false)
			.getResponse();
	},
}

export const handler = Alexa.SkillBuilders.custom()
	.addRequestHandlers(
		LaunchRequest,
		NextLevelHandler,
		RepeatQuestionHandler,
		StopHandler,
		AnswerHandler, // This must be the last non-generic input handler, because Alexa maps other phrases to our slot values.
		// HelpHandler,
		// ExitHandler,
		SessionEndedRequestHandler,
		FallbackHandler,
	)
	.addErrorHandlers(ErrorHandler)
	.lambda()

process.on('unhandledRejection', error => {
	console.error('unhandledRejection', error);
});

const isTestEnvironment = (): boolean => process.env.JEST_WORKER_ID !== undefined
