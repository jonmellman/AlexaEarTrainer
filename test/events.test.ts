// @ts-nocheck
// nocheck because we intentionally test with objects not conforming to types in source code
import { handler } from '../src'

describe('events', () => {
	const testHandler = (input) => {
		return new Promise((resolve, reject) => {
			handler(input, {}, (err, result) => {
				if (err) {
					return reject(err)
				}

				return resolve(result)
			})
		})
	}

	it('user-initiated exit event', async () => {
		const input = {
			"version": "1.0",
			"session": {
				"new": false,
				"sessionId": "amzn1.echo-api.session.7535ef4c-4cbd-42ca-9f45-7ccd98db7f36",
				"application": {
					"applicationId": "amzn1.ask.skill.c8898de4-05ca-44f6-8710-24a56e97933d"
				},
				"attributes": {
					"currentRound": {
						"roundNumber": 1,
						"referenceNote": 60,
						"targetNote": 65,
						"intervalDistance": 5
					},
					"level": 1,
					"stats": []
				},
				"user": {
					"userId": "amzn1.ask.account.AGN5YIY47D3D4YOSYFDGIOEAX4RZUUADVSUNRA2KB3RHX6VHES37TP4KQSNOAJVWTZFUIZOGCE547G3JH5P4G4IBA4JXWNZWLZGJTWAFS4DMITP56UXIQDFJFQEXR3PKFGWAMYDLD4ATC77FMN4EDXOR5X7BDX4ZQ2JXD5PRRIMLH3KLMMVOXWO2WVDLXXE753WRCGS7MY6CBGQ"
				}
			},
			"context": {
				"System": {
					"application": {
						"applicationId": "amzn1.ask.skill.c8898de4-05ca-44f6-8710-24a56e97933d"
					},
					"user": {
						"userId": "amzn1.ask.account.AGN5YIY47D3D4YOSYFDGIOEAX4RZUUADVSUNRA2KB3RHX6VHES37TP4KQSNOAJVWTZFUIZOGCE547G3JH5P4G4IBA4JXWNZWLZGJTWAFS4DMITP56UXIQDFJFQEXR3PKFGWAMYDLD4ATC77FMN4EDXOR5X7BDX4ZQ2JXD5PRRIMLH3KLMMVOXWO2WVDLXXE753WRCGS7MY6CBGQ"
					},
					"device": {
						"deviceId": "amzn1.ask.device.AGPUGKPHH5FCF4QZEWHQZOMN4CZ5DA4FJ2D44VQLQN6VAMXMBHVZUDJBRD7WI2VATOBYXSBE76F3UH36AWOT4X7WBB2WSGEB6BASOD4OLITOJERFC7YG5R7MD7EQFEIP7LDK2PGPHJFT74OTJ4FXAPPCZCHNH2TFEHPGZUXBWV7O3QP7AVA2W",
						"supportedInterfaces": {}
					},
					"apiEndpoint": "https://api.amazonalexa.com",
					"apiAccessToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjEifQ.eyJhdWQiOiJodHRwczovL2FwaS5hbWF6b25hbGV4YS5jb20iLCJpc3MiOiJBbGV4YVNraWxsS2l0Iiwic3ViIjoiYW16bjEuYXNrLnNraWxsLmM4ODk4ZGU0LTA1Y2EtNDRmNi04NzEwLTI0YTU2ZTk3OTMzZCIsImV4cCI6MTU3NzMzMjQ3NywiaWF0IjoxNTc3MzMyMTc3LCJuYmYiOjE1NzczMzIxNzcsInByaXZhdGVDbGFpbXMiOnsiY29udGV4dCI6IkFBQUFBQUFBQUFCQkYzOThjV3pYdGtVblRsY3BHT0s3S2dFQUFBQUFBQUNwTjNXL24xYU4zY0dxUUhnYWpwemhEYXBIS0VnelJpSnF1U0pmK3ZTQ2J1N2dYcGVmd3YzY3Z4RjJUcDlId0FVYUVPODN1SHovS2xzRDdQdXFKOUVCakZRRXR1QlN4b1E4bnF3SC9kaXhJM2xRWjJVK2tiZ3NhZlpRbXVJRlBUbEFRNGtpTy8xTERLQ1JzQTN1T0dkaTBxUDdyTWNmcFRUQzZuTkROM3l6T2duSC9RWG4vTVVZUjNQK000cU9HUDV1MlZBRUFLSVVxbVQ4K2gyNXVnZ25tZTdIMVRMUEM2Wjg5c1VOM09YVEh0VWZWUTV0cDV4TVpRdEI0Z1ArV3lwQjNkYjZzRm5oRlNFd0Zpc04xcVRSWlA0VzdwcTc4OGl3MmF6bFo4Tk1XbTJ2MEhpRFAxa2pjWk5qZFd0eVRtNHFDVWRlTGI3Z1VvVkxQSU1qTE9ITHR4OHFPNzlxaVdTSWpBQmFNRWVheXAyNGJZZnZ1dnZvMERVekJPcTFVY20vZm5lZFFodklTNUdRIiwiY29uc2VudFRva2VuIjpudWxsLCJkZXZpY2VJZCI6ImFtem4xLmFzay5kZXZpY2UuQUdQVUdLUEhINUZDRjRRWkVXSFFaT01ONENaNURBNEZKMkQ0NFZRTFFONlZBTVhNQkhWWlVESkJSRDdXSTJWQVRPQllYU0JFNzZGM1VIMzZBV09UNFg3V0JCMldTR0VCNkJBU09ENE9MSVRPSkVSRkM3WUc1UjdNRDdFUUZFSVA3TERLMlBHUEhKRlQ3NE9USjRGWEFQUENaQ0hOSDJURkVIUEdaVVhCV1Y3TzNRUDdBVkEyVyIsInVzZXJJZCI6ImFtem4xLmFzay5hY2NvdW50LkFHTjVZSVk0N0QzRDRZT1NZRkRHSU9FQVg0UlpVVUFEVlNVTlJBMktCM1JIWDZWSEVTMzdUUDRLUVNOT0FKVldUWkZVSVpPR0NFNTQ3RzNKSDVQNEc0SUJBNEpYV05aV0xaR0pUV0FGUzRETUlUUDU2VVhJUURGSkZRRVhSM1BLRkdXQU1ZRExENEFUQzc3Rk1ONEVEWE9SNVg3QkRYNFpRMkpYRDVQUlJJTUxIM0tMTU1WT1hXTzJXVkRMWFhFNzUzV1JDR1M3TVk2Q0JHUSJ9fQ.SQ9FXK3D6VnIUlm062zIw3TamOsafmLiyL6YCoTUTfs4D5zKQqjIf8MU3fxyh6L3c1XW5ZCjW9nW8wyC6WbvltO1KefkzgstdVLyeP4Q7FoIWNB8bRfByCWGsCX1aVZQHzTvrRkUnk_LMYN5d4JkZPeMrlmLNanP4gunSJOHDXn22WCHZfup8-t-gsNkEeo02ThXib0CuPraCn3GsIogjP3SmUMwGK4Pcvfi_TD9PUs6_49Ui2PkTueqCOHklnPX2smEWHowG4wsmOszdfQ8Bb-ElxVeVsOGwckHm43UIvDlQqoTM5iTIDOZ_yjOZRhcj4SQbj8jXeBOuATG1gm60A"
				},
				"Viewport": {
					"experiences": [
						{
							"arcMinuteWidth": 246,
							"arcMinuteHeight": 144,
							"canRotate": false,
							"canResize": false
						}
					],
					"shape": "RECTANGLE",
					"pixelWidth": 1024,
					"pixelHeight": 600,
					"dpi": 160,
					"currentPixelWidth": 1024,
					"currentPixelHeight": 600,
					"touch": [
						"SINGLE"
					],
					"video": {
						"codecs": [
							"H_264_42",
							"H_264_41"
						]
					}
				},
				"Viewports": [
					{
						"type": "APL",
						"id": "main",
						"shape": "RECTANGLE",
						"dpi": 160,
						"presentationType": "STANDARD",
						"canRotate": false,
						"configuration": {
							"current": {
								"video": {
									"codecs": [
										"H_264_42",
										"H_264_41"
									]
								},
								"size": {
									"type": "DISCRETE",
									"pixelWidth": 1024,
									"pixelHeight": 600
								}
							}
						}
					}
				]
			},
			"request": {
				"type": "SessionEndedRequest",
				"requestId": "amzn1.echo-api.request.e2caa15e-3e22-4ee8-83cc-e6dcaa7e2fed",
				"timestamp": "2019-12-26T03:49:37Z",
				"locale": "en-US",
				"reason": "USER_INITIATED"
			}
		}

		const result = await testHandler(input)
		expect(result).toMatchSnapshot()
	})

	it('interrupt alexa with cancel', async () => {
		const input = {
			"version": "1.0",
			"session": {
				"new": false,
				"sessionId": "amzn1.echo-api.session.16979802-2d72-453c-832d-4c1e040b9594",
				"application": {
					"applicationId": "amzn1.ask.skill.c8898de4-05ca-44f6-8710-24a56e97933d"
				},
				"attributes": {
					"currentRound": {
						"roundNumber": 1,
						"referenceNote": 60,
						"targetNote": 62,
						"intervalDistance": 2
					},
					"level": 1,
					"stats": []
				},
				"user": {
					"userId": "amzn1.ask.account.AGN5YIY47D3D4YOSYFDGIOEAX4RZUUADVSUNRA2KB3RHX6VHES37TP4KQSNOAJVWTZFUIZOGCE547G3JH5P4G4IBA4JXWNZWLZGJTWAFS4DMITP56UXIQDFJFQEXR3PKFGWAMYDLD4ATC77FMN4EDXOR5X7BDX4ZQ2JXD5PRRIMLH3KLMMVOXWO2WVDLXXE753WRCGS7MY6CBGQ"
				}
			},
			"context": {
				"System": {
					"application": {
						"applicationId": "amzn1.ask.skill.c8898de4-05ca-44f6-8710-24a56e97933d"
					},
					"user": {
						"userId": "amzn1.ask.account.AGN5YIY47D3D4YOSYFDGIOEAX4RZUUADVSUNRA2KB3RHX6VHES37TP4KQSNOAJVWTZFUIZOGCE547G3JH5P4G4IBA4JXWNZWLZGJTWAFS4DMITP56UXIQDFJFQEXR3PKFGWAMYDLD4ATC77FMN4EDXOR5X7BDX4ZQ2JXD5PRRIMLH3KLMMVOXWO2WVDLXXE753WRCGS7MY6CBGQ"
					},
					"device": {
						"deviceId": "amzn1.ask.device.AGPUGKPHH5FCF4QZEWHQZOMN4CZ5DA4FJ2D44VQLQN6VAMXMBHVZUDJBRD7WI2VATOBYXSBE76F3UH36AWOT4X7WBB2WSGEB6BASOD4OLITOJERFC7YG5R7MD7EQFEIP7LDK2PGPHJFT74OTJ4FXAPPCZCHNH2TFEHPGZUXBWV7O3QP7AVA2W",
						"supportedInterfaces": {}
					},
					"apiEndpoint": "https://api.amazonalexa.com",
					"apiAccessToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjEifQ.eyJhdWQiOiJodHRwczovL2FwaS5hbWF6b25hbGV4YS5jb20iLCJpc3MiOiJBbGV4YVNraWxsS2l0Iiwic3ViIjoiYW16bjEuYXNrLnNraWxsLmM4ODk4ZGU0LTA1Y2EtNDRmNi04NzEwLTI0YTU2ZTk3OTMzZCIsImV4cCI6MTU3NzMzNDUxMCwiaWF0IjoxNTc3MzM0MjEwLCJuYmYiOjE1NzczMzQyMTAsInByaXZhdGVDbGFpbXMiOnsiY29udGV4dCI6IkFBQUFBQUFBQUFCQkYzOThjV3pYdGtVblRsY3BHT0s3S2dFQUFBQUFBQUE1SitaNmtrN1dGSzIzWHIwRHlmcGVUbCtPQU1sRy94akpOSU5QTUxhMjczSWVKeEd0bWVvMmd1NU0xUW9MeXlTZm5kV2xWU2R1TC9zekNObmt4eGp4eGtCanlSR0tjYWYzM0hzeWo2Q0MzZnp2aWNON1cvaTJCb05lUXQwb1NOM2JXUlVEV0xEM1NmdmNwWHd4Sk5hQ2ZUREZNcWw2KytRb1JxV2czaHdaWlRSWjV3dzZLTklnK0g5Y2E2TFBBc25Fd2ZKTTNqRHpjVExkd2xDY3Z1YUR4cnRlQ091Y3Z2bk01djQxSllweVNFYVlxamFKa242ZGh6bSs2NUtDNURIdVljNmNQK1JoMXFiV2pEWHI1UTd2Zzc0UTYrMTFEQ0NsalhZbnMvZUxvMzJZZW9hd1JHY1dYZVlyRXNNajRGcWR0bkR5N3ZGdUFBSno4WTdVeVNTaHVoa3k5SXdoY1pVUU1ZazBwdFo5SVhtYlhUdmlQTGF6OS9yNHZDVEVxb2I4a01EcXRaMGw2V0NtIiwiY29uc2VudFRva2VuIjpudWxsLCJkZXZpY2VJZCI6ImFtem4xLmFzay5kZXZpY2UuQUdQVUdLUEhINUZDRjRRWkVXSFFaT01ONENaNURBNEZKMkQ0NFZRTFFONlZBTVhNQkhWWlVESkJSRDdXSTJWQVRPQllYU0JFNzZGM1VIMzZBV09UNFg3V0JCMldTR0VCNkJBU09ENE9MSVRPSkVSRkM3WUc1UjdNRDdFUUZFSVA3TERLMlBHUEhKRlQ3NE9USjRGWEFQUENaQ0hOSDJURkVIUEdaVVhCV1Y3TzNRUDdBVkEyVyIsInVzZXJJZCI6ImFtem4xLmFzay5hY2NvdW50LkFHTjVZSVk0N0QzRDRZT1NZRkRHSU9FQVg0UlpVVUFEVlNVTlJBMktCM1JIWDZWSEVTMzdUUDRLUVNOT0FKVldUWkZVSVpPR0NFNTQ3RzNKSDVQNEc0SUJBNEpYV05aV0xaR0pUV0FGUzRETUlUUDU2VVhJUURGSkZRRVhSM1BLRkdXQU1ZRExENEFUQzc3Rk1ONEVEWE9SNVg3QkRYNFpRMkpYRDVQUlJJTUxIM0tMTU1WT1hXTzJXVkRMWFhFNzUzV1JDR1M3TVk2Q0JHUSJ9fQ.HxVKdez504QdfqpB8BRoRfIkOf26Eg85EwnwGqUvUSpJcu51NR79-prq_1V5F9GmOUUwnqHw8QGhG548McecKKmwrelfHYuUes3eZMSeV07p95GknSBVJ-62JLbP-TYsVmBJY80P1cCFYderoWdkOmjdu6ltZa_igZF4aXtJMr6VsQGcA9e7DSR-Tsp59PYmLOzi9XIqNu9_wn1TOrKCvKKVlQtzjaaPCCJeoQY7iWmnZ4dyfZXchCl0ZAtg04kNty26WwyXto0W-ndIamsZUG090iBeWdDDjpnWY15NaWS-7Z83rDqMd2qNMT69aBkyAsV9qC_bQmQvIpnQ8iqU2g"
				},
				"Viewport": {
					"experiences": [
						{
							"arcMinuteWidth": 246,
							"arcMinuteHeight": 144,
							"canRotate": false,
							"canResize": false
						}
					],
					"shape": "RECTANGLE",
					"pixelWidth": 1024,
					"pixelHeight": 600,
					"dpi": 160,
					"currentPixelWidth": 1024,
					"currentPixelHeight": 600,
					"touch": [
						"SINGLE"
					],
					"video": {
						"codecs": [
							"H_264_42",
							"H_264_41"
						]
					}
				},
				"Viewports": [
					{
						"type": "APL",
						"id": "main",
						"shape": "RECTANGLE",
						"dpi": 160,
						"presentationType": "STANDARD",
						"canRotate": false,
						"configuration": {
							"current": {
								"video": {
									"codecs": [
										"H_264_42",
										"H_264_41"
									]
								},
								"size": {
									"type": "DISCRETE",
									"pixelWidth": 1024,
									"pixelHeight": 600
								}
							}
						}
					}
				]
			},
			"request": {
				"type": "IntentRequest",
				"requestId": "amzn1.echo-api.request.b96dfcd9-f149-4df9-8f1f-4add5bcd96fd",
				"timestamp": "2019-12-26T04:23:30Z",
				"locale": "en-US",
				"intent": {
					"name": "AMAZON.CancelIntent",
					"confirmationStatus": "NONE"
				}
			}
		}

		const result = await testHandler(input)
		expect(result).toMatchSnapshot()
	})
})
