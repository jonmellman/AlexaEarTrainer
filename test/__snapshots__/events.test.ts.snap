// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events interrupt alexa with cancel 1`] = `
Object {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak>Thanks for playing!</speak>",
      "type": "SSML",
    },
    "shouldEndSession": true,
  },
  "sessionAttributes": Object {
    "currentRound": Object {
      "intervalDistance": 2,
      "referenceNote": 60,
      "roundNumber": 1,
      "targetNote": 62,
    },
    "level": 1,
    "stats": Array [],
  },
  "userAgent": "ask-node/2.7.0 Node/v12.13.0",
  "version": "1.0",
}
`;

exports[`events user-initiated exit event 1`] = `
Object {
  "response": Object {},
  "sessionAttributes": Object {
    "currentRound": Object {
      "intervalDistance": 5,
      "referenceNote": 60,
      "roundNumber": 1,
      "targetNote": 65,
    },
    "level": 1,
    "stats": Array [],
  },
  "userAgent": "ask-node/2.7.0 Node/v12.13.0",
  "version": "1.0",
}
`;
